src/fdc3/dotnet/DesktopAgent.Client/src/MorganStanley.ComposeUI.Fdc3.DesktopAgent.Client/Infrastructure/DesktopAgentClient.cs

```diff
@@ -27,35 +29,94 @@ namespace MorganStanley.ComposeUI.Fdc3.DesktopAgent.Client.Infrastructure;
     private readonly string _instanceId;
-    private readonly JsonSerializerOptions _jsonSerializerOptions = SerializerOptionsHelper.JsonSerializerOptions;
+    private readonly IChannelFactory _channelFactory;
+    private readonly IMetadataClient _metadataClient;
+    private readonly ConcurrentDictionary<IListener, Func<string, ChannelType, CancellationToken, ValueTask>> _contextListenersWithSubscriptionLastContextHandlingActions = new();
+
+    private IChannel? _currentChannel;
+    private readonly SemaphoreSlim _currentChannelLock = new(1, 1);
+
+    private readonly ConcurrentDictionary<string, IChannel> _userChannels = new();
 
     public DesktopAgentClient(
         IMessaging messaging,
-        ILogger<DesktopAgentClient>? logger = null)
+        ILoggerFactory? loggerFactory = null)
     {
         _messaging = messaging;
-        _logger = logger ?? NullLogger<DesktopAgentClient>.Instance;
+        _loggerFactory = loggerFactory ?? NullLoggerFactory.Instance;
+        _logger = _loggerFactory.CreateLogger<DesktopAgentClient>();
+
+        _appId = Environment.GetEnvironmentVariable(nameof(AppIdentifier.AppId)) ?? throw ThrowHelper.MissingAppId(string.Empty);
+        _instanceId = Environment.GetEnvironmentVariable(nameof(AppIdentifier.InstanceId)) ?? throw ThrowHelper.MissingInstanceId(_appId, string.Empty);
 
-        _appId = Environment.GetEnvironmentVariable(nameof(AppIdentifier.AppId)) ?? throw new Fdc3DesktopAgentException("AppId for the app cannot be retrieved!");
-        _instanceId = Environment.GetEnvironmentVariable(nameof(AppIdentifier.InstanceId)) ?? throw new Fdc3DesktopAgentException("InstanceId defined by the ModuleLoader/StartupAction for the app cannot be retrieved!");
+        _channelFactory = new ChannelFactory(_messaging, _instanceId, _loggerFactory);
+        _metadataClient = new MetadataClient(_appId, _instanceId, _messaging, _loggerFactory.CreateLogger<MetadataClient>());
     }
 
-    public Task<IListener> AddContextListener<T>(string? contextType, ContextHandler<T> handler) where T : IContext
+    //TODO: AddContextListener should be revisited when the Open is being implemented as the first context that should be handled is the context which is passed through the fdc3.open call.
+    public async Task<IListener> AddContextListener<T>(string? contextType, ContextHandler<T> handler) where T : IContext
     {
-        throw new NotImplementedException();
+        ContextListener<T> listener = null;
+        try
+        {
+            await _currentChannelLock.WaitAsync().ConfigureAwait(false);
+
+            listener = await _channelFactory.CreateContextListener(handler, _currentChannel, contextType);
+            if (!_contextListenersWithSubscriptionLastContextHandlingActions.TryAdd(
+                listener, 
+                async(channelId, channelType, cancellationToken) => 
+                { 
+                    await listener.SubscribeAsync(channelId, channelType, cancellationToken); 
```

Comment:
Subscribing first, then handling the last context can cause a race condition between an incoming context and the existing one - possibly handling the incoming one twice